### ISSUE TITLE
[1.7] Logs do not get filtered by service properly when new containers start

### ISSUE THREAD

**Issue by JacobHayes** [docker/compose#3361]
When monitoring logs from a service (`docker-compose logs -f service1`), container starts from a different service will be shown in the log output. The feature in 1.7 to add new containers to exiting log outputs is nice, but should still be limited to containers of the same service when one is specified (ie: only when the scale changes).

To recreate:
```yaml
version: '2'
services:
    service1:
        image: alpine
        command: sh -c 'while true; do echo "service1"; sleep 1; done;'
    service2:
        image: alpine
        command: sh -c 'while true; do echo "service2"; sleep 1; done;'
```

```
docker-compose up -d service1 && \
eval '(sleep 5 && docker-compose up -d service2) &' && \
docker-compose logs -f service1
```

After five seconds, you will start to see output from both services, even though we just requested logs for `service1`.

**Comment by dnephin** [comment]
Not sure why this is broken. We should be handling this because we filter events to just those that match the services names. So there should be no way to start watching a container for a service that isn't listed.

**Comment by aronahl** [comment]
LogPrinter's watch_events method starts watching events for any unknown container, regardless of name.

This seems to fix it (only cursory testing):

```diff
diff --git a/compose/cli/log_printer.py b/compose/cli/log_printer.py
index b48462f..90b3c6c 100644
--- a/compose/cli/log_printer.py
+++ b/compose/cli/log_printer.py
@@ -62,13 +62,15 @@ class LogPrinter(object):
                  event_stream,
                  output=sys.stdout,
                  cascade_stop=False,
-                 log_args=None):
+                 log_args=None,
+                 services=None):
         self.containers = containers
         self.presenters = presenters
         self.event_stream = event_stream
         self.output = utils.get_output_stream(output)
         self.cascade_stop = cascade_stop
         self.log_args = log_args or {}
+        self.services = services
 
     def run(self):
         if not self.containers:
@@ -81,7 +83,8 @@ class LogPrinter(object):
             thread_map,
             self.event_stream,
             self.presenters,
-            thread_args))
+            thread_args,
+            self.services))
 
         for line in consume_queue(queue, self.cascade_stop):
             remove_stopped_threads(thread_map)
@@ -186,7 +189,7 @@ def start_producer_thread(thread_args):
     producer.start()
 
 
-def watch_events(thread_map, event_stream, presenters, thread_args):
+def watch_events(thread_map, event_stream, presenters, thread_args, services):
     for event in event_stream:
         if event['action'] == 'stop':
             thread_map.pop(event['id'], None)
@@ -200,10 +203,11 @@ def watch_events(thread_map, event_stream, presenters, thread_args):
             # Container was stopped and started, we need a new thread
             thread_map.pop(event['id'], None)
 
-        thread_map[event['id']] = build_thread(
-            event['container'],
-            next(presenters),
-            *thread_args)
+        if event['service'] in services:
+            thread_map[event['id']] = build_thread(
+                event['container'],
+                next(presenters),
+                *thread_args)
 
 
 def consume_queue(queue, cascade_stop):
diff --git a/compose/cli/main.py b/compose/cli/main.py
index 8348b8c..a381f8d 100644
--- a/compose/cli/main.py
+++ b/compose/cli/main.py
@@ -411,7 +411,8 @@ class TopLevelCommand(object):
             self.project,
             containers,
             options['--no-color'],
-            log_args).run()
+            log_args,
+            services=options['SERVICE']).run()
 
     def pause(self, options):
         """
@@ -736,7 +737,8 @@ class TopLevelCommand(object):
                 options['--no-color'],
                 {'follow': True},
                 cascade_stop,
-                event_stream=self.project.events(service_names=service_names))
+                event_stream=self.project.events(service_names=service_names),
+                services=options['SERVICE'])
             print("Attaching to", list_containers(log_printer.containers))
             log_printer.run()
 
@@ -890,13 +892,15 @@ def log_printer_from_project(
     log_args,
     cascade_stop=False,
     event_stream=None,
+    services = None
 ):
     return LogPrinter(
         containers,
         build_log_presenters(project.service_names, monochrome),
         event_stream or project.events(),
         cascade_stop=cascade_stop,
-        log_args=log_args)
+        log_args=log_args,
+        services=services)
 
 
 def filter_containers_to_service_names(containers, service_names):

```

**Comment by dnephin** [comment]
@aronahl the `event_stream` passed to `LogPrinter` should already be filtering events by the service name:

* https://github.com/docker/compose/blob/7ad7eb71ca704a43bcdcd4a923f58fdc6e49b62a/compose/project.py#L356-L358
* https://github.com/docker/compose/blob/7ad7eb71ca704a43bcdcd4a923f58fdc6e49b62a/compose/cli/main.py#L754

I don't think adding another arg to all of these functions is the correct solution. We need to find out why the existing filtering isn't working as expected.

**Comment by aronahl** [comment]
Oh.

Then it just needs:

```diff
diff --git a/compose/cli/main.py b/compose/cli/main.py
index 8348b8c..4f90964 100644
--- a/compose/cli/main.py
+++ b/compose/cli/main.py
@@ -411,7 +411,8 @@ class TopLevelCommand(object):
             self.project,
             containers,
             options['--no-color'],
-            log_args).run()
+            log_args,
+            event_stream=self.project.events(service_names=options['SERVICE'])).run()

     def pause(self, options):
```

**Comment by dnephin** [comment]
Oh, yes, that's the bug. We filter by service name on `up`, but not on `logs`. I was looking at the wrong function,

Thanks! would you mind submitting a PR for this fix?

**Comment by aronahl** [comment]
Will do.

**Comment by shin-** [comment]
Fixed by #3390
